### PR TITLE
make export namespace aware

### PR DESCRIPTION
**Pull request by NamanJain8**
This PR makes export and backup/restore to work. 
**Export**
Export will be namespace-aware with each RDF containing the info about the namespace. 
In NQuads, I have kept that in the label.

RDF format:
```
<0x1> <dgraph.type> "dgraph.type.cors"^^<xs:string> <0x0> .
<0x2> <person_name> "alice berry" <0x0> .
...
<0x9> <person_name> "bob berry" <0x12> .
```
JSON format:
```
{"uid":"0x7532","namespace":"0x0","person_name":"bob berry"},
```
Schema Format:
```
[0x0] <dgraph.cors>:[string] @index(exact) @upsert . 
[0x0] <dgraph.type>:[string] @index(exact) . 
[0x0] <person_name>:default . 
...
[0x0] <dgraph.graphql.schema_created_at>:datetime . 
[0x12] <person_name>:default . 
...
[0x13] <dgraph.cors>:[string] @index(exact) @upsert . 
[0x13] <dgraph.type>:[string] @index(exact) . 
[0x13] <person_name>:default . 
...
[0x14] <dgraph.graphql.p_sha256hash>:string @index(exact) . 
[0x14] <dgraph.graphql.schema_history>:string . 
[0x14] <dgraph.graphql.schema_created_at>:datetime . 
[0x0] type <dgraph.graphql> {
	dgraph.graphql.schema
	dgraph.graphql.xid
}
[0x0] type <dgraph.type.cors> {
	dgraph.cors
}
...
```

**Backup**
We can't figure out if the backups taken on an older version (`v20.11`), or on `20.03`. 
Either we ask the user to specify that and trust him, else we add a field in the manifest
about the version the backup was taken on.
The backup taken on `v20.11`, when restored, goes into the default namespace.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7377)

### CODE REVIEW

File: ee/backup/run.go [hunk 1]
```diff
@@ -294,6 +296,8 @@ func initExportBackup() {
 		"The folder to which export the backups.")
 	flag.StringVarP(&opt.format, "format", "f", "rdf",
 		"The format of the export output. Accepts a value of either rdf or json")
+	flag.Uint64VarP(&opt.namespace, "namespace", "n", math.MaxUint64, "The namespace for which the"+
+		"backup has to be exported. By default, it will export all namespaces.")
```
Comment (NamanJain8) [inline ee/backup/run.go:300]: I haven't verified the functioning of `export_backup` command.

File: worker/backup_common.go [hunk 2]
```diff
@@ -54,6 +54,8 @@ type Manifest struct {
 	// backup gets assigned the next available number. Used to verify the integrity
 	// of the data during a restore.
 	BackupNum uint64 `json:"backup_num"`
+	// Version specifies the Dgraph version, the backup was taken on.
+	Version string `json:"version"`
```
Comment (NamanJain8) [inline worker/backup_common.go:58]: This is the DgraphVersion. This helps in figuring out which version the backup was taken on.
I think we should use a number as we may later need >= type of check later on. We can parse it later as well.

File: graphql/admin/export.go [hunk 3]
```diff
@@ -50,8 +52,11 @@ func resolveExport(ctx context.Context, m schema.Mutation) (*resolve.Resolved, b
 		}
 	}
 
+	// TODO: Check if that namespace exist and the relevant permissions.
```
Comment (jarifibrahim) [inline graphql/admin/export.go:55]: Add name next to todo.

File: graphql/admin/admin.go [hunk 4]
```diff
@@ -208,6 +208,11 @@ const (
 		"""
 		format: String
 
+		"""
+		Namespace for the export, if no value is given then it exports all namespaces.
```
Comment (jarifibrahim) [inline graphql/admin/admin.go:212]: Don't make this mandatory. We can find the namespace from the jwt token.

File: protos/pb.proto [hunk 5]
```diff
@@ -649,6 +649,8 @@ message ExportRequest {
 	string secret_key = 7;
 	string session_token = 8;
 	bool anonymous = 9;
+
+	uint64 namespace = 10;
```
Comment (jarifibrahim) [inline protos/pb.proto:653]: You don't need this. The token will have this information.

Comment (jarifibrahim) [reply protos/pb.proto:653]: Check ctx for token, if not found, export default namespace.

How would we export all the namespaces?
 -> Use the guardian of the galaxy token for this.

File: schema/state.go [hunk 6]
```diff
@@ -137,6 +155,16 @@ func isNameBegin(r rune) bool {
 	}
 }
 
+// isNumberBegin returns true if the rune is an number.
```
Comment (jarifibrahim) [inline schema/state.go:158]: isNumber

File: schema/state.go [hunk 7]
```diff
@@ -137,6 +155,16 @@ func isNameBegin(r rune) bool {
...
+func isNumber(r rune) bool {
+	switch {
+	case r >= '0' && r <= '9':
```
Comment (jarifibrahim) [inline schema/state.go:161]: return ` r >= '0' && r <= '9':`

File: worker/backup_common.go (hunk 2 above)
Comment (jarifibrahim) [reply worker/backup_common.go:59]: Use numbers instead. 2103, 2011, etc

File: worker/export.go [hunk 8]
```diff
@@ -142,7 +145,7 @@ func (e *exporter) toJSON() (*bpb.KVList, error) {
 	// Leaving it simple for now.
 
 	continuing := false
-	mapStart := fmt.Sprintf("  {\"uid\":"+uidFmtStrJson, e.uid)
+	mapStart := fmt.Sprintf("  {\"uid\":"+uidFmtStrJson+`,"namespace":"0x%x"`, e.uid, e.namespace)
```
Comment (jarifibrahim) [inline worker/export.go:148]: Add a small example in the PR description

File: worker/export.go [hunk 9]
```diff
@@ -578,6 +593,11 @@ func exportInternal(ctx context.Context, in *pb.ExportRequest, db *badger.DB,
 	}
 
 	stream := db.NewStreamAt(in.ReadTs)
+	stream.Prefix = []byte{x.DefaultPrefix}
+	if in.Namespace != math.MaxUint64 {
```
Comment (jarifibrahim) [inline worker/export.go:597]: Get this from token.

File: worker/export.go [hunk 10]
```diff
@@ -597,22 +617,13 @@ func exportInternal(ctx context.Context, in *pb.ExportRequest, db *badger.DB,
 			return false
 		}
 
-		// _predicate_ is deprecated but leaving this here so that users with a
-		// binary with version >= 1.1 can export data from a version < 1.1 without
```
Comment (jarifibrahim) [inline worker/export.go:601]: Bring this back.

File: worker/export.go [hunk 11]
```diff
@@ -578,6 +593,11 @@ func exportInternal(ctx context.Context, in *pb.ExportRequest, db *badger.DB,
 	}
 
 	stream := db.NewStreamAt(in.ReadTs)
```
Comment (jarifibrahim) [inline worker/export.go:595]: Add a comment that this stream is just for the data

**Review by jarifibrahim — COMMENTED** [review]
Got some comments. I still have to review some code.

File: worker/export.go [hunk 12]
```diff
@@ -791,6 +772,77 @@ func exportInternal(ctx context.Context, in *pb.ExportRequest, db *badger.DB,
...
+				var update pb.TypeUpdate
+				err := item.Value(func(val []byte) error {
+					return update.Unmarshal(val)
+				})
+				if err != nil {
+					// Let's not propagate this error. We just log this and continue onwards.
+					glog.Errorf("Unable to unmarshal type: %+v. Err=%v\n", pk, err)
+					return nil
+				}
+				kvs, err = toType(pk.Attr, update)
```
Comment (jarifibrahim) [inline worker/export.go:839]: This error is not being handled.

Comment (jarifibrahim) [reply worker/export.go:839]: the same for `toSchema` as well

File: edgraph/server.go [hunk 13]
```diff
@@ -114,6 +114,7 @@ func (s *Server) CreateNamespace(ctx context.Context, namespace uint64) error {
 	glog.Info("Creating namespace", namespace)
 	m := &pb.Mutations{StartTs: worker.State.GetTimestamp(false)}
 	m.Schema = schema.InitialSchema(namespace)
+	m.Types = schema.InitialTypes(namespace)
```
[the remark this answers is not in the document]
Comment (NamanJain8) [reply edgraph/server.go:117]: This is working fine for my case.

File: worker/export.go (hunk 12 above)
Comment (NamanJain8) [reply worker/export.go:839]: They don't return an error. So removed it from the return values of the function.

File: ee/backup/run.go [hunk 14]
```diff
@@ -323,5 +327,5 @@ func runExportBackup() error {
 	}
 
 	exporter := worker.BackupExporter{}
-	return exporter.ExportBackup(opt.location, opt.destination, opt.format, opt.key)
+	return exporter.ExportBackup(opt.location, opt.destination, opt.format, opt.namespace, opt.key)
```
[the remark this answers is not in the document]
Comment (NamanJain8) [reply ee/backup/run.go:330]: We will need to export the options. Should we?

File: edgraph/server.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jarifibrahim) [reply edgraph/server.go:117]: But don't you need to wait for the indexing to complete?

File: ee/backup/run.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jarifibrahim) [reply ee/backup/run.go:330]: You'll have to create a new option type in file_handler.go file. If you create the new type here (or expose opt), you'll run into cyclic dependency.

File: schema/parse.go [hunk 15]
```diff
@@ -433,8 +456,12 @@ func isTypeDeclaration(item lex.Item, it *lex.ItemIterator) bool {
 	return true
 }
 
-// Parse parses a schema string and returns the schema representation for it.
-func Parse(s string) (*ParsedSchema, error) {
+// parse parses a schema string and returns the schema representation for it.
+func parse(s string, namespace uint64) (*ParsedSchema, error) {
```
Comment (NamanJain8) [inline schema/parse.go:460]: Examples:
Schema 1:
```
[ns1] name: string .
[ns2] age: string .
city: string .
```

parse(schema, 0) --> All the schema fields go to namespace 0.
parse(schema, x) --> All the schema fields go to namespace x.
parse(schema, math.MaxUint64) --> name (ns1), age(ns2), city(namespace 0)

This function can be modified if we want to enforce that the schema should contain a namespace. Though not a big issue, but I think allowing schema without namespace will be helpful with data loading.
If we want the user to give either consistent schema (all fields should have namespace or none should), then we can make this code section simpler.

**Review by manishrjain — APPROVED** [review]
:lgtm: Remove the version stuff by using an extra field for namespace in backup key proto. Fix the lexer to spit and parse hex namespaces, but still work with int namespaces.

Rest of it looks good.

> Reviewed 17 of 38 files at r3, 7 of 13 files at r4, 3 of 4 files at r5, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-:-MSdQs9FEpy1oWsOwdap:b-dgunha)* status: all files reviewed, 31 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[edgraph/server.go, line 117 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSSp24iLiqVX0IDWFs8-r3-117:-MSdL_BP7P5u_dDSUNUh:bbeunfl) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/edgraph/server.go#L117)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

But don't you need to wait for the indexing to complete?
</blockquote></details>

I think types need to be replicated to all the groups, if I'm not wrong. But, we could change the behavior by having types lie only on group 1, and having those streamed out to other groups, like we do with ACL.xn
___
*[ee/backup/run.go, line 318 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdLrZL5EVY8BZs5p5P:-MSdLrZL5EVY8BZs5p5Q:bnn2yia) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/ee/backup/run.go#L318)):*
> ```Go
> 	flag.StringVarP(&opt.format, "format", "f", "rdf",
> 		"The format of the export output. Accepts a value of either rdf or json")
> 	flag.Uint64VarP(&opt.namespace, "namespace", "n", math.MaxUint64, "The namespace for which the"+
> ```

Backup can only backup every namespace. Not a specific namespace. So, no need for this flag.
___
*[ee/backup/run.go, line 330 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdLxff4WiL3dWS6_of:-MSdLxff4WiL3dWS6_og:bu2u8yk) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/ee/backup/run.go#L330)):*
> ```Go
> 
> 	exporter := worker.BackupExporter{}
> 	return exporter.ExportBackup(opt.location, opt.destination, opt.format, opt.namespace, opt.key)
> ```

Instead of having 5 arguments, just send opt itself.
___
*[posting/index.go, line 579 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdMNFV8X-DrBmeB_Hz:-MSdMNFV8X-DrBmeB_I-:b-necci9) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/posting/index.go#L579)):*
> ```Go
> 		WithEncryptionKey(x.WorkerConfig.EncryptionKey).
> 		WithLoggingLevel(badger.WARNING).
> 		WithNamespaceOffset(1)
> ```

Don't need it here?
___
*[schema/parse.go, line 417 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdMc2qChtJQs1s0nT5:-MSdMc2qChtJQs1s0nT6:bgxx1r2) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/schema/parse.go#L417)):*
> ```Go
> 		return 0, errors.Errorf("Unable to peek: %v", err)
> 	}
> 	if nextItems[0].Typ != itemNumber || nextItems[1].Typ != itemRightSquare {
> ```

can itemNumber identify hex?

"0xabc" should be identified as a number. Use hex.
___
*[schema/parse.go, line 461 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdN3jICuCH59wTgLIZ:-MSdN3jICuCH59wTgLI_:bgwvwwx) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/schema/parse.go#L461)):*
> ```Go
> // parse parses a schema string and returns the schema representation for it.
> func parse(s string, namespace uint64) (*ParsedSchema, error) {
> 	defaultNs := x.DefaultNamespace
> ```

Add a comment for the roles that namespace and defaultNs play. It's not clear why are there two different variables.
___
*[schema/state.go, line 113 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdO2XTDOkmyeh-I_gk:-MSdO2XTDOkmyeh-I_gl:b-y525b0) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/schema/state.go#L113)):*
> ```Go
> }
> 
> func lexNumber(l *lex.Lexer) lex.StateFn {
> ```

Make this also parse hex.
___
*[worker/export.go, line 290 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOWQgCe2kh8Y0Jjyu:-MSdOWQgCe2kh8Y0Jjyv:b-yhku12) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/export.go#L290)):*
> ```Go
> func toSchema(attr string, update *pb.SchemaUpdate) *bpb.KV {
> 	// bytes.Buffer never returns error for any of the writes. So, we don't need to check them.
> 	ns, attr := x.ParseNamespaceAttr(attr)
> ```

We should have a separate field in the backup key proto, which contains the namespace. It would be zero for older backups. And it would have a legit value for newer backups. So, it would work naturally.
___
*[worker/export.go, line 813 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdPxDm2N5q9WcSRfEI:-MSdPxDm2N5q9WcSRfEJ:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/export.go#L813)):*
> ```Go
> 			case x.ByteSchema:
> 				if !skipZero {
> 					if servesTablet, err := groups().ServesTablet(pk.Attr); err != nil || !servesTablet {
> ```

100 chars.
___
*[worker/online_restore_ee.go, line 240 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOgIe1E6iV0O-RlGk:-MSdOgIe1E6iV0O-RlGl:b-vc87jj) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/online_restore_ee.go#L240)):*
> ```Go
> 
> 	// Version is 0 if the backup was taken on an old version (v20.11).
> 	if lastManifest.Version == 0 {
> ```

Remove all the version stuff.x
___
*[worker/sort.go, line 61 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOlbZ9u124rcMoQdI:-MSdOlbZ9u124rcMoQdJ:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/sort.go#L61)):*
> ```Go
> 		return &emptySortResult, err
> 	} else if gid == 0 {
> 		return &emptySortResult, errors.Errorf("Cannot sort by unknown attribute %s", x.ParseAttr(q.Order[0].Attr))
> ```

100 chars.
___
*[worker/sort.go, line 65 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOnVO9aSKoFkfN5Dv:-MSdOnVO9aSKoFkfN5Dw:b-yq3bu8) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/sort.go#L65)):*
> ```Go
> 
> 	if span := otrace.FromContext(ctx); span != nil {
> 		span.Annotatef(nil, "worker.SortOverNetwork. Attr: %s. Group: %d", x.ParseAttr(q.Order[0].Attr), gid)
> ```

and here.
___
*[x/keys.go, line 253 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdP-uo8Wu3dUl_s0_L:-MSdP-uo8Wu3dUl_s0_M:bz20xsw) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/x/keys.go#L253)):*
> ```Go
> }
> 
> func DefaultSchemaKey(attr string) []byte {
> ```

can't we say

`SchemaKey(attr string, ns uint64)`?
___
*[x/keys.go, line 258 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdPaH12tDNsL95kLnr:-MSdPaH12tDNsL95kLns:bljgeti) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/x/keys.go#L258)):*
> ```Go
> }
> 
> func DefaultTypeKey(attr string) []byte {
> ```

Move them to test util or something, if possible.
___
*[x/x.go, line 134 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdPi5c0oOkTNHBNtAM:-MSdPi5d3D7HQGXzfIK3:b6f48os) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/x/x.go#L134)):*
> ```Go
> 	DgraphCostHeader = "Dgraph-TouchedUids"
> 
> 	DgraphVersion = 2103
> ```

don't need all this.

**Review by jarifibrahim — APPROVED** [review]
:lgtm:

> Reviewed 2 of 38 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-:-MShY5n3EZYpgdeWNZ0B:bnfp4nl)* status: 12 of 27 files reviewed, 31 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[graphql/admin/export.go, line 99 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShS6KmCa7tCHkBTV5d:-MShS6KmCa7tCHkBTV5e:b-oejkqg) ([raw file](https://github.com/dgraph-io/dgraph/blob/7793eec2e72cc669ce2aa7eaedd707474e481476/graphql/admin/export.go#L99)):*
> ```Go
> 	err = json.Unmarshal(inputByts, &input)
> 
> 	// Export everything if namespace is not specified.
> ```

Add a todo to fix this and use jwt later
___
*[schema/parse.go, line 287 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShSM-a7SSn-cdKLUNl:-MShSM-a7SSn-cdKLUNm:b-dqyxyv) ([raw file](https://github.com/dgraph-io/dgraph/blob/7793eec2e72cc669ce2aa7eaedd707474e481476/schema/parse.go#L287)):*
> ```Go
> 				seen[tokenizer.Name()] = true
> 			} else {
> 				return errors.Errorf("Duplicate tokenizers present for attr %s", x.ParseAttr(schema.Predicate))
> ```

100 chars
___
*[worker/export.go, line 865 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShWDs94y8DH_OxUvwX:-MShWDs94y8DH_OxUvwY:bg3erax) ([raw file](https://github.com/dgraph-io/dgraph/blob/7793eec2e72cc669ce2aa7eaedd707474e481476/worker/export.go#L865)):*
> ```Go
> 		return nil, err
> 	}
> 	if err := writeSchema(x.ByteType); err != nil {
> ```

call it writePrefix
___
*[x/keys.go, line 321 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShXgBa7DW8qLJHaExA:-MShXgBb1_AV7uNNh-W9:b-6nbld7) ([raw file](https://github.com/dgraph-io/dgraph/blob/e8745c891594349e275d06612b4de92df5da587c/x/keys.go#L321)):*
> ```Go
> 
> // SkipSchema returns the first key after all the schema keys.
> func (p ParsedKey) SkipSchema() []byte {
> ```

Remove these in a separate PR.

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7377)* status: 11 of 27 files reviewed, 30 unresolved discussions (waiting on @ahsanbarkati, @danielmai, @jarifibrahim, @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[edgraph/server.go, line 289 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSSp24iLiqVX0IDWFs8-r3-289:-MShSi2sAOaBmeKl00t3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/edgraph/server.go#L289)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Remove this.
</blockquote></details>

Done.
___
*[edgraph/server.go, line 314 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSSp24iLiqVX0IDWFs8-r3-314:-MShSiUMCCOYEo-iBm_W:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/edgraph/server.go#L314)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Remove this as well.
</blockquote></details>

Done.
___
*[graphql/admin/admin.go, line 212 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MS7uUrgDWbHEwDjdvrR-r3-212:-MShZd4766xNwbY0Wlfi:bqhegxh) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/graphql/admin/admin.go#L212)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Don't make this mandatory. We can find the namespace from the jwt token.
</blockquote></details>

Will be done once ACL is there.
___
*[graphql/admin/export.go, line 55 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MS7uUrgDWbHEwDjdvrS-r3-55:-MShZgEAA-fKMwrrnigT:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/graphql/admin/export.go#L55)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add name next to todo.
</blockquote></details>

Done.
___
*[graphql/admin/export.go, line 99 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShS6KmCa7tCHkBTV5d:-MShZhxuDxW7qwVqU9yJ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7793eec2e72cc669ce2aa7eaedd707474e481476/graphql/admin/export.go#L99)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add a todo to fix this and use jwt later
</blockquote></details>

Done.
___
*[posting/index.go, line 579 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdMNFV8X-DrBmeB_Hz:-MShZj-Z5oPcgBVobnqY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/posting/index.go#L579)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need it here?
</blockquote></details>

Done.
___
*[protos/pb.proto, line 653 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MS7uUrgDWbHEwDjdvrT-r3-653:-MShZkbRC3GRox9QUI-E:b-ii0vg) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/protos/pb.proto#L653)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Check ctx for token, if not found, export default namespace.

How would we export all the namespaces?
 -> Use the guardian of the galaxy token for this.
</blockquote></details>

Will be done with ACL.
___
*[schema/parse.go, line 417 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdMc2qChtJQs1s0nT5:-MShZpT-CTdMB4jXjeN2:bbqksi) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/schema/parse.go#L417)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

can itemNumber identify hex?

"0xabc" should be identified as a number. Use hex.
</blockquote></details>

Done. It can lex hexadecimal as well as integer now.
___
*[schema/parse.go, line 461 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdN3jICuCH59wTgLIZ:-MSh_-Nf7aOn_aRBSkwK:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/schema/parse.go#L461)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add a comment for the roles that namespace and defaultNs play. It's not clear why are there two different variables.
</blockquote></details>

Done.
___
*[schema/parse.go, line 287 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShSM-a7SSn-cdKLUNl:-MShZnIa5RNZezW5epnu:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7793eec2e72cc669ce2aa7eaedd707474e481476/schema/parse.go#L287)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

100 chars
</blockquote></details>

Done.
___
*[schema/state.go, line 113 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdO2XTDOkmyeh-I_gk:-MSh_17m7IHifGzs2VfG:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/schema/state.go#L113)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Make this also parse hex.
</blockquote></details>

Done.
___
*[worker/export.go, line 595 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MS7uUrgDWbHEwDjdvrV-r3-595:-MSh_BbW0-SHJo0rFghi:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/worker/export.go#L595)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Add a comment that this stream is just for the data
</blockquote></details>

Done.
___
*[worker/export.go, line 597 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MS7uUrgDWbHEwDjdvrV-r3-597:-MSh_Crq2qxJ9-v_V-Dc:b-hukn99) ([raw file](https://github.com/dgraph-io/dgraph/blob/dc8ad6ab9d59962dd314ac94215a966858abbd60/worker/export.go#L597)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

Get this from token.
</blockquote></details>

Added TODO.
___
*[worker/export.go, line 829 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MS7uUrgDWbHEwDjdvrV-r4-829:-MSh_Hg4EnhzE4bW7SFW:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fc45a44df1a0b7fc71d036dae5af20b3cd1c6050/worker/export.go#L829)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

They don't return an error. So removed it from the return values of the function.
</blockquote></details>

Done.
___
*[worker/export.go, line 290 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOWQgCe2kh8Y0Jjyu:-MSh_A1CCRI2uSU90GRm:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/export.go#L290)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

We should have a separate field in the backup key proto, which contains the namespace. It would be zero for older backups. And it would have a legit value for newer backups. So, it would work naturally.
</blockquote></details>

Done.
___
*[worker/export.go, line 813 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdPxDm2N5q9WcSRfEI:-MSh_FrB8DYJrkTHBCJc:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/export.go#L813)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.
___
*[worker/export.go, line 865 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MShWDs94y8DH_OxUvwX:-MSh_IXsEpkX4oLUS0yA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7793eec2e72cc669ce2aa7eaedd707474e481476/worker/export.go#L865)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

call it writePrefix
</blockquote></details>

Done.
___
*[worker/online_restore_ee.go, line 240 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOgIe1E6iV0O-RlGk:-MSh_kAK5H4sLsciO-dz:bmb41p6) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/online_restore_ee.go#L240)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove all the version stuff.x
</blockquote></details>

This is needed because for the predicates stored in manifest.Group does not have info regarding the namespace. It will not be prefixed with a namespace in case of older backups.
___
*[worker/sort.go, line 61 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOlbZ9u124rcMoQdI:-MSh_Lcm6cwY6HGqCk1J:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/sort.go#L61)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.
___
*[worker/sort.go, line 65 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdOnVO9aSKoFkfN5Dv:-MSh_MNl4dWaXqEfhquj:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/worker/sort.go#L65)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

and here.
</blockquote></details>

Done.
___
*[x/keys.go, line 258 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7377#-MSdPaH12tDNsL95kLnr:-MSh_Nyo-igPdK-e4QOB:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9a21af1fc067425a1f7bdf6c179c25647789bd0d/x/keys.go#L258)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Move them to test util or something, if possible.
</blockquote></details>

Done.

**Pull request merged**